File: library/src/test/java/me/proxer/library/api/HttpsEnforcingInterceptorTest.java
package me.proxer.library.api;

import me.proxer.library.ProxerTest;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.mockwebserver.MockResponse;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class HttpsEnforcingInterceptorTest extends ProxerTest {

    @Test
    void testHttpsUpgradeWeb() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());

        client.newCall(new Request.Builder().url("http://proxer.me").build()).execute();

        assertThat(server.takeRequest().getRequestUrl().isHttps()).isEqualTo(true);
    }

    @Test
    void testHttpsUpgradeCdn() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());

        client.newCall(new Request.Builder().url("http://cdn.proxer.me").build()).execute();

        assertThat(server.takeRequest().getRequestUrl().isHttps()).isEqualTo(true);
    }

    @Test
    void testHttpsUpgradeManga() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());

        client.newCall(new Request.Builder().url("http://manga1.proxer.me").build()).execute();

        assertThat(server.takeRequest().getRequestUrl().isHttps()).isEqualTo(true);
    }

    @Test
    void testHttpsUpgradeStreamHtml() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());

        client.newCall(new Request.Builder().url("http://stream.proxer.me").build()).execute();

        assertThat(server.takeRequest().getRequestUrl().isHttps()).isEqualTo(true);
    }

    @Test
    void testHttpsUpgradeStreamAlternative() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());

        client.newCall(new Request.Builder().url("http://s1-ps.proxer.me").build()).execute();

        assertThat(server.takeRequest().getRequestUrl().isHttps()).isEqualTo(true);
    }

    @Test
    void testHttpsUntouched() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());

        client.newCall(new Request.Builder().url("https://proxer.me").build()).execute();

        assertThat(server.takeRequest().getRequestUrl().isHttps()).isEqualTo(true);
    }

    @Test
    void testOtherHostThrows() {
        final HttpsEnforcingInterceptor interceptor = new HttpsEnforcingInterceptor();
        final Interceptor.Chain chain = mock(Interceptor.Chain.class);

        when(chain.request()).thenReturn(new Request.Builder().url("https://example.com").build());

        assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> interceptor.intercept(chain));
    }
}

File: library/src/test/kotlin/me/proxer/library/api/HttpsEnforcingInterceptorTest.kt
package me.proxer.library.api

import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.Test
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.notNull
import org.mockito.Mockito.verify

class HttpsEnforcingInterceptorTest {

    private val interceptor = HttpsEnforcingInterceptor()
    private val chain = mock(Interceptor.Chain::class.java)

    @Test
    fun testHttpsUpgradeWeb() {
        val requestCaptor = ArgumentCaptor.forClass(Request::class.java)

        `when`(chain.request()).thenReturn(Request.Builder().url("http://proxer.me").build())
        `when`(chain.proceed(notNull())).thenReturn(mock(Response::class.java))

        interceptor.intercept(chain)

        verify(chain).proceed(requestCaptor.capture())
        assertThat(requestCaptor.value.url().toString()).isEqualTo("https://proxer.me/")
    }

    @Test
    fun testHttpsUpgradeCdn() {
        val requestCaptor = ArgumentCaptor.forClass(Request::class.java)

        `when`(chain.request()).thenReturn(Request.Builder().url("http://cdn.proxer.me").build())
        `when`(chain.proceed(notNull())).thenReturn(mock(Response::class.java))

        interceptor.intercept(chain)

        verify(chain).proceed(requestCaptor.capture())
        assertThat(requestCaptor.value.url().toString()).isEqualTo("https://cdn.proxer.me/")
    }

    @Test
    fun testHttpsUpgradeManga() {
        val requestCaptor = ArgumentCaptor.forClass(Request::class.java)

        `when`(chain.request()).thenReturn(Request.Builder().url("http://manga1.proxer.me").build())
        `when`(chain.proceed(notNull())).thenReturn(mock(Response::class.java))

        interceptor.intercept(chain)

        verify(chain).proceed(requestCaptor.capture())
        assertThat(requestCaptor.value.url().toString()).isEqualTo("https://manga1.proxer.me/")
    }

    @Test
    fun testHttpsUpgradeStreamHtml() {
        val requestCaptor = ArgumentCaptor.forClass(Request::class.java)

        `when`(chain.request()).thenReturn(Request.Builder().url("http://stream.proxer.me").build())
        `when`(chain.proceed(notNull())).thenReturn(mock(Response::class.java))

        interceptor.intercept(chain)

        verify(chain).proceed(requestCaptor.capture())
        assertThat(requestCaptor.value.url().toString()).isEqualTo("https://stream.proxer.me/")
    }

    @Test
    fun testHttpsUpgradeStreamAlternative() {
        val requestCaptor = ArgumentCaptor.forClass(Request::class.java)

        `when`(chain.request()).thenReturn(Request.Builder().url("http://s1-ps.proxer.me").build())
        `when`(chain.proceed(notNull())).thenReturn(mock(Response::class.java))

        interceptor.intercept(chain)

        verify(chain).proceed(requestCaptor.capture())
        assertThat(requestCaptor.value.url().toString()).isEqualTo("https://s1-ps.proxer.me/")
    }

    @Test
    fun testHttpsUntouched() {
        val requestCaptor = ArgumentCaptor.forClass(Request::class.java)

        `when`(chain.request()).thenReturn(Request.Builder().url("https://proxer.me").build())
        `when`(chain.proceed(notNull())).thenReturn(mock(Response::class.java))

        interceptor.intercept(chain)

        verify(chain).proceed(requestCaptor.capture())
        assertThat(requestCaptor.value.url().toString()).isEqualTo("https://proxer.me/")
    }

    @Test
    fun testOtherHostThrows() {
        `when`(chain.request()).thenReturn(Request.Builder().url("https://example.com").build())
        `when`(chain.proceed(notNull())).thenReturn(mock(Response::class.java))

        assertThatExceptionOfType(IllegalArgumentException::class.java).isThrownBy { interceptor.intercept(chain) }
    }
}
